public class main
{
    static op void p2(int) throws java.io.EOFException;

    static process run
    {
	  while(true)
	  {
	    inni void p2(int i) throws java.io.EOFException
	    {
		    reply handler new Handler2();
		throw new java.io.EOFException();
	    }
	  }
    }


    public static void main(String [] args)
    {
	    send p2(3) handler new Handler();
	    try
	    {
		call p2(4);
	    }
	    catch (java.io.EOFException e)
	    {
		System.out.println("Caught through call to p2 with try/catch");
	    }
    }
}

class Handler implements edu.ucdavis.jr.Handler
{
    public handler op void method(java.io.EOFException e);
    public handler void method(java.io.EOFException e)
    {
	System.out.println("Caught java.io.EOFException " +
				"through handler object");
    }
}

class Handler2 implements edu.ucdavis.jr.Handler
{
    public handler op void method(java.io.EOFException e);
    public handler void method(java.io.EOFException e)
    {
	System.out.println("Caught java.io.EOFException " +
				"through Handler2 object");
    }
}

